File: ios/TripCore/Sources/TripCore/TripPackage.swift
import Foundation

/// The contract the web planner publishes and this app consumes.
///
/// Mirrors `src/lib/tripPackage.js`. Everything here exists on both sides, so
/// changing a field means changing it twice — the golden fixtures at
/// `fixtures/trips/` fail loudly when only one side moves.
///
/// Decoding is deliberately tolerant of *unknown* fields (Swift's synthesised
/// `Codable` ignores them) so a newer planner can add data without bricking an
/// older build, but strict about the ones it needs.
public struct TripPackage: Codable, Sendable, Equatable {
    public static let expectedFormat = "ulpacker-trip-package"
    public static let supportedSchemaVersion = 1
    public static let supportedHashAlgorithm = "fnv1a64"

    public let format: String
    public let schemaVersion: Int
    public let hashAlgorithm: String
    public let tripId: String
    public let revision: Int
    public let publishedAt: String
    public let trip: Trip
    public let plannedRoute: PlannedRoute
    public let checkpoints: [Checkpoint]
    public let itinerary: [Day]
    public let extraDays: [ExtraDay]
    public let navigationDefaults: NavigationDefaults
    public let contentHash: String

    public struct Trip: Codable, Sendable, Equatable {
        public let name: String
        public let description: String
        public let startName: String
        public let finishName: String
        public let loop: Bool
        public let startDayNumber: Int
    }

    public struct PlannedRoute: Codable, Sendable, Equatable {
        public let segments: [Segment]
        public let stats: Stats
    }

    /// A run of consecutive points. Two segments mean a real gap in the route
    /// (a shuttle, a ferry), not a rendering artefact — the planner already
    /// fuses ways that merely touch.
    public struct Segment: Codable, Sendable, Equatable {
        public let points: [TrackPoint]
    }

    public struct Stats: Codable, Sendable, Equatable {
        public let distanceM: Int
        public let ascentM: Int?
        public let descentM: Int?
        public let minEle: Int?
        public let maxEle: Int?
        public let elevationCoverage: Double
        public let pointCount: Int
        public let segmentCount: Int
    }

    /// Ordered by `routeDistanceM` — guaranteed by the builder, not by the
    /// caller, so the "next checkpoint" cursor can rely on it.
    public struct Checkpoint: Codable, Sendable, Equatable {
        public let id: String
        public let name: String
        public let note: String
        public let kind: String
        public let routeDistanceM: Int
        public let lat: Double
        public let lng: Double
        public let ele: Int?
    }

    public struct Day: Codable, Sendable, Equatable {
        public let index: Int
        public let startRouteM: Int
        public let endRouteM: Int
        public let distanceM: Int
        public let startBoundary: String
        public let startName: String
        public let endBoundary: String
        public let endName: String
        public let ascentM: Int?
        public let descentM: Int?
        public let note: String
    }

    /// A day with no geometry — travel, rest, shuttle. Shown in the list,
    /// never navigated.
    public struct ExtraDay: Codable, Sendable, Equatable {
        public let id: String
        public let before: String
        public let title: String
        public let note: String
    }

    public struct NavigationDefaults: Codable, Sendable, Equatable {
        public let offRouteEnterM: Int
        public let offRouteExitM: Int
    }
}

/// A point as it appears on the wire: `[lat, lng]` or `[lat, lng, ele]`.
///
/// JSON arrays rather than objects, because at 8.5k points per route the key
/// names would triple the file for no gain. The custom coding keeps that
/// compactness without leaking `[Double]` into the rest of the code.
public struct TrackPoint: Codable, Sendable, Equatable {
    public let lat: Double
    public let lng: Double
    public let ele: Int?

    public init(lat: Double, lng: Double, ele: Int?) {
        self.lat = lat
        self.lng = lng
        self.ele = ele
    }

    public init(from decoder: Decoder) throws {
        var container = try decoder.unkeyedContainer()
        lat = try container.decode(Double.self)
        lng = try container.decode(Double.self)
        ele = container.isAtEnd ? nil : try container.decodeIfPresent(Int.self)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.unkeyedContainer()
        try container.encode(lat)
        try container.encode(lng)
        if let ele { try container.encode(ele) }
    }
}

// MARK: - Validation

public enum TripPackageError: Error, Equatable, CustomStringConvertible {
    case wrongFormat(String)
    case unsupportedSchemaVersion(Int)
    case unsupportedHashAlgorithm(String)
    case emptyRoute
    case coordinateOutOfRange(lat: Double, lng: Double)
    case contentHashMismatch(expected: String, actual: String)

    public var description: String {
        switch self {
        case .wrongFormat(let f):
            return "Not a trip package (format \"\(f)\")."
        case .unsupportedSchemaVersion(let v):
            return "Unsupported schemaVersion \(v); this build reads \(TripPackage.supportedSchemaVersion)."
        case .unsupportedHashAlgorithm(let a):
            return "Unsupported hashAlgorithm \"\(a)\"."
        case .emptyRoute:
            return "The package has no route to follow."
        case .coordinateOutOfRange(let lat, let lng):
            return "Coordinate out of range (\(lat), \(lng))."
        case .contentHashMismatch(let expected, let actual):
            return "contentHash mismatch (file says \(expected), content hashes to \(actual))."
        }
    }
}

extension TripPackage {
    /// Recompute the content hash of a package's raw bytes.
    ///
    /// Takes `Data` rather than a decoded value because the hash describes the
    /// file: fields this build doesn't model still have to count, or a newer
    /// planner's additions would silently stop being covered.
    public static func contentHash(of data: Data) throws -> String {
        // `publishedAt` changes on every export and `contentHash` cannot cover
        // itself, so both stay out — re-exporting an unchanged trip must yield
        // the same hash.
        let canonical = try CanonicalJSON.canonicalize(
            data: data,
            omittingTopLevelKeys: ["contentHash", "publishedAt"]
        )
        return FNV1a.hash64(canonical)
    }

    /// Decode and fully verify a package, including its content hash.
    ///
    /// Verification is not optional on this path: a package arrives by AirDrop
    /// or Drive and is then trusted for navigation for days, so a silently
    /// truncated file has to fail here rather than halfway up a mountain.
    public static func decode(from data: Data) throws -> TripPackage {
        let package = try JSONDecoder().decode(TripPackage.self, from: data)
        try package.validateStructure()
        let actual = try contentHash(of: data)
        guard actual == package.contentHash else {
            throw TripPackageError.contentHashMismatch(expected: package.contentHash, actual: actual)
        }
        return package
    }

    /// Structural checks only. The hash is verified in `decode(from:)`, which
    /// is the only place the original bytes are still available.
    public func validateStructure() throws {
        guard format == Self.expectedFormat else { throw TripPackageError.wrongFormat(format) }
        guard schemaVersion == Self.supportedSchemaVersion else {
            throw TripPackageError.unsupportedSchemaVersion(schemaVersion)
        }
        guard hashAlgorithm == Self.supportedHashAlgorithm else {
            throw TripPackageError.unsupportedHashAlgorithm(hashAlgorithm)
        }
        guard !plannedRoute.segments.isEmpty,
              plannedRoute.segments.allSatisfy({ $0.points.count >= 2 }) else {
            throw TripPackageError.emptyRoute
        }
        for segment in plannedRoute.segments {
            for point in segment.points {
                guard (-90...90).contains(point.lat), (-180...180).contains(point.lng) else {
                    throw TripPackageError.coordinateOutOfRange(lat: point.lat, lng: point.lng)
                }
            }
        }
    }
}

File: src/lib/tripPackage.js
// TripPackage v1 — the contract between the web planner and the iOS companion.
//
// The web app owns trips; the companion only ever reads them, navigates, and
// records an activity alongside. So this is deliberately NOT the app's internal
// trip shape: it carries what navigation needs (route geometry, checkpoints,
// itinerary, thresholds) and drops what it doesn't (gear, packs, cover images,
// editor-only state). Anything added here has to be implemented twice — once in
// JS, once in Swift — so the bar for adding a field is high.
//
// A package is immutable for a given `tripId + revision`. `contentHash` is the
// content identity: it covers everything except `publishedAt` and the hash
// field itself, so re-exporting an unchanged trip yields the same hash.

import { buildCumulatives, buildDays, buildTrackStats } from "./trail.js";

export const TRIP_PACKAGE_FORMAT = "ulpacker-trip-package";
export const TRIP_PACKAGE_VERSION = 1;

// Not a security hash — this detects "did the content change", nothing more.
// FNV-1a was picked over SHA-256 because it stays synchronous (no WebCrypto
// promise in an export handler) and is ~15 lines in Swift as well as here.
// The algorithm is named in the payload so it can be swapped without ambiguity.
export const HASH_ALGORITHM = "fnv1a64";

// Off-route defaults. 75 m is a provisional value to be field-calibrated (see
// MOBILE_PLAN §3.2): a static 200 m threshold only fires ~3 minutes after the
// walker left the route. Exit sits below enter so the state machine has
// hysteresis instead of oscillating on GPS noise.
export const DEFAULT_OFF_ROUTE_ENTER_M = 75;
export const DEFAULT_OFF_ROUTE_EXIT_M = 40;

// Matches COORD_DECIMALS in trail.js (~1.1 m). Coordinates are rounded before
// hashing so the hash can't depend on float noise from a re-import.
const COORD_DP = 5;
// Number of decimals used when a non-integer is serialised for hashing. Both
// languages must format identically, so the rule is fixed, not "shortest form".
const HASH_DP = 6;

function roundTo(value, dp) {
  const f = 10 ** dp;
  return Math.round(Number(value) * f) / f;
}

// --- Canonical serialisation (must be reimplemented identically in Swift) ---
//
// Object keys sorted; numbers emitted by an explicit rule (integers as digits,
// everything else at fixed HASH_DP decimals) because JS and Swift disagree on
// the "shortest round-trip" representation of a float. Undefined and null
// members are dropped so an absent field and an explicitly-null one hash alike.

function canonicalNumber(n) {
  if (!Number.isFinite(n)) throw new TypeError("Cannot canonicalise a non-finite number");
  return Number.isInteger(n) ? String(n) : n.toFixed(HASH_DP);
}

export function canonicalJson(value) {
  if (value === null || value === undefined) return "null";
  if (typeof value === "number") return canonicalNumber(value);
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "string") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(",")}]`;
  if (typeof value === "object") {
    const parts = [];
    for (const key of Object.keys(value).sort()) {
      const v = value[key];
      if (v === undefined || v === null) continue;
      parts.push(`${JSON.stringify(key)}:${canonicalJson(v)}`);
    }
    return `{${parts.join(",")}}`;
  }
  throw new TypeError(`Cannot canonicalise ${typeof value}`);
}

// 64-bit FNV-1a over the UTF-8 bytes, as 16 lowercase hex digits.
export function fnv1a64(text) {
  const MASK = 0xffffffffffffffffn;
  const PRIME = 0x100000001b3n;
  let hash = 0xcbf29ce484222325n;
  for (const byte of new TextEncoder().encode(text)) {
    hash = ((hash ^ BigInt(byte)) * PRIME) & MASK;
  }
  return hash.toString(16).padStart(16, "0");
}

// The hash covers the package minus the two fields that must not affect
// identity: the hash itself, and the timestamp of this particular export.
export function computeContentHash(pkg) {
  const { contentHash, publishedAt, ...rest } = pkg;
  return fnv1a64(canonicalJson(rest));
}

// --- Build -----------------------------------------------------------------

function packageSegments(segments) {
  const out = [];
  for (const seg of segments || []) {
    const points = [];
    for (const p of seg?.points || []) {
      if (!Array.isArray(p)) continue;
      const lat = Number(p[0]);
      const lng = Number(p[1]);
      if (!Number.isFinite(lat) || !Number.isFinite(lng)) continue;
      const ele = Number.isFinite(Number(p[2])) ? Math.round(Number(p[2])) : null;
      points.push(ele == null ? [roundTo(lat, COORD_DP), roundTo(lng, COORD_DP)] : [roundTo(lat, COORD_DP), roundTo(lng, COORD_DP), ele]);
    }
    if (points.length >= 2) out.push({ points });
  }
  return out;
}

// Checkpoints keep only what navigation reads. The internal anchor
// (segmentIndex/alongSegmentM/offsetM/source*) is dropped: routeDistanceM plus
// the coordinates are enough, and re-deriving the rest is the companion's job.
//
// Sorted by route distance here rather than trusting the caller: the app keeps
// trip.checkpoints ordered via normalizeTrips(), but the companion reads these
// as a sequence ("next checkpoint"), so the ordering has to be a property of
// the contract instead of an invariant maintained on the other side of it.
function packageCheckpoints(checkpoints) {
  return (checkpoints || [])
    .map((cp) => {
      const a = cp?.anchor;
      if (!a || !Number.isFinite(a.lat) || !Number.isFinite(a.lng)) return null;
      return {
        id: cp.id,
        name: cp.name || "",
        note: cp.note || "",
        kind: cp.kind || "poi",
        routeDistanceM: Math.round(Number(a.routeDistanceM) || 0),
        lat: roundTo(a.lat, COORD_DP),
        lng: roundTo(a.lng, COORD_DP),
        ele: Number.isFinite(a.ele) ? Math.round(a.ele) : null
      };
    })
    .filter(Boolean)
    .sort((x, y) => x.routeDistanceM - y.routeDistanceM);
}

// buildDays() returns `{ days, warnings }` normally but a bare `[]` when the
// track is empty (an inconsistency in trail.js we work around rather than
// change, since the web UI depends on the current behaviour).
function packageItinerary(trip, segments) {
  if (!segments.length) return [];
  const cumulatives = buildCumulatives(segments);
  const built = buildDays({ checkpoints: trip.checkpoints || [], segments, cumulatives });
  const days = Array.isArray(built) ? [] : built.days || [];
  const notes = trip.dayNotes && typeof trip.dayNotes === "object" ? trip.dayNotes : {};
  return days.map((day) => ({
    index: day.index,
    startRouteM: Math.round(day.startRouteM),
    endRouteM: Math.round(day.endRouteM),
    distanceM: Math.round(day.distanceM),
    startBoundary: day.startBoundary,
    startName: day.startName || "",
    endBoundary: day.endBoundary,
    endName: day.endName || "",
    ascentM: day.ascentM == null ? null : Math.round(day.ascentM),
    descentM: day.descentM == null ? null : Math.round(day.descentM),
    note: typeof notes[day.startBoundary] === "string" ? notes[day.startBoundary] : ""
  }));
}

// Off-track itinerary days (travel, rest, shuttle). They carry no geometry, so
// the companion shows them in the day list but never navigates them.
function packageExtraDays(extraDays) {
  return (extraDays || [])
    .filter((d) => d && typeof d === "object")
    .map((d) => ({
      id: d.id,
      before: d.before || "finish",
      title: d.title || "",
      note: typeof d.note === "string" ? d.note : ""
    }));
}

/**
 * Build a TripPackage v1 from the app's internal trip + track.
 *
 * Returns null when the trip has no usable route — a package without geometry
 * is useless to a navigator, so callers should surface that to the user rather
 * than shipping an empty package.
 *
 * `revision` is caller-supplied and defaults to 1: trips do not yet persist a
 * revision (adding one means widening the trip schema in storage.js, which is
 * deliberately out of scope for the contract freeze). `contentHash` already
 * answers "is this different"; `revision` answers "is this newer", and wiring
 * it to persistent state is a follow-up.
 */
export function buildTripPackage(trip, track, options = {}) {
  if (!trip || typeof trip !== "object") return null;
  const segments = packageSegments(track?.segments);
  if (!segments.length) return null;

  const revision = Number.isFinite(Number(options.revision)) ? Math.max(1, Math.trunc(Number(options.revision))) : 1;
  const publishedAt = options.publishedAt || new Date().toISOString();
  const stats = buildTrackStats(segments);

  const pkg = {
    format: TRIP_PACKAGE_FORMAT,
    schemaVersion: TRIP_PACKAGE_VERSION,
    hashAlgorithm: HASH_ALGORITHM,
    tripId: trip.id,
    revision,
    publishedAt,
    trip: {
      name: trip.name || "Untitled trip",
      description: trip.description || "",
      startName: trip.startName || "",
      finishName: trip.finishName || "",
      loop: Boolean(trip.loop),
      startDayNumber: trip.startDayNumber === 0 ? 0 : 1
    },
    plannedRoute: {
      segments,
      stats: {
        distanceM: Math.round(stats.distanceM),
        ascentM: stats.ascentM == null ? null : Math.round(stats.ascentM),
        descentM: stats.descentM == null ? null : Math.round(stats.descentM),
        minEle: stats.minEle == null ? null : Math.round(stats.minEle),
        maxEle: stats.maxEle == null ? null : Math.round(stats.maxEle),
        elevationCoverage: roundTo(stats.elevationCoverage, 3),
        pointCount: segments.reduce((n, s) => n + s.points.length, 0),
        segmentCount: segments.length
      }
    },
    checkpoints: packageCheckpoints(trip.checkpoints),
    itinerary: packageItinerary(trip, segments),
    extraDays: packageExtraDays(trip.extraDays),
    navigationDefaults: {
      offRouteEnterM: Number.isFinite(Number(options.offRouteEnterM))
        ? Math.max(1, Math.round(Number(options.offRouteEnterM)))
        : DEFAULT_OFF_ROUTE_ENTER_M,
      offRouteExitM: Number.isFinite(Number(options.offRouteExitM))
        ? Math.max(1, Math.round(Number(options.offRouteExitM)))
        : DEFAULT_OFF_ROUTE_EXIT_M
    }
  };

  pkg.contentHash = computeContentHash(pkg);
  return pkg;
}

// --- Validate --------------------------------------------------------------

/**
 * Validate an untrusted TripPackage (a file the companion or a user hands us).
 * Returns { ok, errors } — errors is empty when ok. Structural problems and a
 * mismatched hash are both failures; an unknown newer schemaVersion is also a
 * failure here, because this validator only knows v1.
 */
export function validateTripPackage(raw) {
  const errors = [];
  const fail = (msg) => {
    errors.push(msg);
    return { ok: false, errors };
  };

  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return fail("Not an object.");
  if (raw.format !== TRIP_PACKAGE_FORMAT) return fail(`Wrong format (expected "${TRIP_PACKAGE_FORMAT}").`);
  if (raw.schemaVersion !== TRIP_PACKAGE_VERSION) {
    return fail(`Unsupported schemaVersion ${raw.schemaVersion} (this build reads ${TRIP_PACKAGE_VERSION}).`);
  }
  if (raw.hashAlgorithm !== HASH_ALGORITHM) return fail(`Unsupported hashAlgorithm "${raw.hashAlgorithm}".`);
  if (typeof raw.tripId !== "string" || !raw.tripId) errors.push("Missing tripId.");
  if (!Number.isInteger(raw.revision) || raw.revision < 1) errors.push("revision must be a positive integer.");
  if (typeof raw.publishedAt !== "string" || Number.isNaN(Date.parse(raw.publishedAt))) {
    errors.push("publishedAt must be an ISO timestamp.");
  }
  if (!raw.trip || typeof raw.trip !== "object") errors.push("Missing trip.");

  const segments = raw.plannedRoute?.segments;
  if (!Array.isArray(segments) || segments.length === 0) {
    errors.push("plannedRoute.segments must be a non-empty array.");
  } else {
    let bad = 0;
    for (const seg of segments) {
      if (!Array.isArray(seg?.points) || seg.points.length < 2) {
        bad += 1;
        continue;
      }
      for (const p of seg.points) {
        const lat = Number(p?.[0]);
        const lng = Number(p?.[1]);
        if (!Number.isFinite(lat) || lat < -90 || lat > 90 || !Number.isFinite(lng) || lng < -180 || lng > 180) {
          bad += 1;
          break;
        }
      }
    }
    if (bad) errors.push(`${bad} segment(s) are empty or contain out-of-range coordinates.`);
  }

  if (!Array.isArray(raw.checkpoints)) errors.push("checkpoints must be an array.");
  if (!Array.isArray(raw.itinerary)) errors.push("itinerary must be an array.");

  if (typeof raw.contentHash !== "string" || !raw.contentHash) {
    errors.push("Missing contentHash.");
  } else if (errors.length === 0) {
    // Only worth checking once the shape is sound; a hash mismatch on a
    // malformed package tells the user nothing useful.
    let actual;
    try {
      actual = computeContentHash(raw);
    } catch (e) {
      errors.push(`Could not hash the package: ${e.message}`);
    }
    if (actual && actual !== raw.contentHash) {
      errors.push(`contentHash mismatch (file says ${raw.contentHash}, content hashes to ${actual}).`);
    }
  }

  return { ok: errors.length === 0, errors };
}
